File: Day9/main.js
function moveTrain(board, mov) {
    const headRow = board.findIndex(row => row.includes('@'));
    const headCol = board[headRow].indexOf('@');

    const moves = { 'L': [0, -1], 'R': [0, 1], 'U': [-1, 0], 'D': [1, 0] };
    const [dRow, dCol] = moves[mov];
    const newRow = headRow + dRow;
    const newCol = headCol + dCol;

    const newPos = board[newRow]?.[newCol];
    if (!newPos || newPos === 'o') return 'crash';
    if (newPos === '*') return 'eat';

    board[newRow] = board[newRow].slice(0, newCol) + '@' + board[newRow].slice(newCol + 1);
    board[headRow] = board[headRow].slice(0, headCol) + 'o' + board[headRow].slice(headCol + 1);

    return 'none';
}

File: Day9/main.py
from typing import List, Literal

def move_train(board: List[str], mov: Literal['U', 'D', 'R', 'L']) -> Literal['none', 'crash', 'eat']:
    head_row = next(i for i, row in enumerate(board) if '@' in row)
    head_col = board[head_row].index('@')

    moves = {'L': (0, -1), 'R': (0, 1), 'U': (-1, 0), 'D': (1, 0)}
    d_row, d_col = moves[mov]

    new_row, new_col = head_row + d_row, head_col + d_col

    if not (0 <= new_row < len(board)) or not (0 <= new_col < len(board[new_row])):
        return 'crash'

    if board[new_row][new_col] in ('o', ' '):
        return 'crash'
    if board[new_row][new_col] == '*':
        return 'eat'

    board[new_row] = board[new_row][:new_col] + '@' + board[new_row][new_col + 1:]
    board[head_row] = board[head_row][:head_col] + 'o' + board[head_row][head_col + 1:]

    return 'none'
